src/Core/Indexes/KDTree.cs

```diff
@@ -0,0 +1,724 @@
+                {
+                    return this.leftChild.Point.Y < envelope.MinY
+                                    ? false
+                                    : true;
+                }
+
+                if (this.splitDimension == 3)
+                {
+                    return this.leftChild.Point.Z < envelope.MinZ
+                                    ? false
+                                    : true;
+                }
+
+                return true;
+            }
+
+            /// <summary>
+            /// Determines whether the right subtree should be searched.
+            /// </summary>
+            /// <param name="envelope">The envelope in which the search is performed.</param>
+            /// <returns>A value indicating whether the right subtree of this node has to be searched.</returns>
+            private bool RightChildShouldBeSearched(Envelope envelope)
+            {
+                if (this.rightChild == null)
+                    return false;
+
+                if (this.splitDimension == 1)
+                {
+                    return this.rightChild.Point.X > envelope.MaxX
+                                    ? false
+                                    : true;
+                }
+
+                if (this.splitDimension == 2)
+                {
+                    return this.rightChild.Point.Y > envelope.MaxY
+                                    ? false
+                                    : true;
+                }
+
+                if (this.splitDimension == 3)
+                {
+                    return this.rightChild.Point.Z > envelope.MaxZ
+                                    ? false
+                                    : true;
+                }
+
+                return true;
+            }
+
+            /// <summary>
+            /// Initializes the left subtree of this node.
+            /// </summary>
+            /// <param name="coordinates">The coordinates to be stored in the left subtree of this node.</param>
+            private void InitializeLeftSubTree(List<Coordinate> coordinates)
+            {
+                if (coordinates.Count == 0)
+                {
+                    this.leftChild = null;
+                    return;
+                }
+
+                int nextSplitDimension = this.splitDimension % this.TotalDimensions + 1;
+                Comparison<Coordinate> coordinateComparison = this.GetComparisonForDimension(nextSplitDimension);
+                coordinates.Sort(coordinateComparison);
+
+                Coordinate median = coordinates[coordinates.Count / 2];
+                this.leftChild = new KDTreeNode(median, nextSplitDimension, this.TotalDimensions);
+
+                List<Coordinate> leftCoordinatesOfChild = coordinates.GetRange(0, coordinates.Count / 2);
+                List<Coordinate> rightCoordinatesOfChild = coordinates.GetRange(coordinates.Count / 2 + 1, (int)Math.Ceiling(coordinates.Count / 2.0) - 1);
```

Comment:
I get your point. In that case I recommend clearing the list after the two `GetRange` queries. That way, we won't allocate coordinates multiple times.